### DISCUSSION TITLE
Errors from different version of PyG

### DISCUSSION THREAD

**Discussion by hkim716** [pyg-team/pytorch_geometric#3620] [Q&A] [answered]
Hi Matt, 

I have a GNN model with ```SplineConv``` made by old version of ```torch_geometric```. (1.7.2) And I tried to load this model on another machine, ```2.0.2 torch_geometric``` was installed on. 

I got an error when I load the model on new machine like this. 

```
RuntimeError: Error(s) in loading state_dict for autoencoder:
	Missing key(s) in state_dict: "encoder.down_path.0.lin.weight", "encoder.down_path.1.lin.weight", "encoder.down_path.2.lin.weight", "encoder.down_path.3.lin.weight", "encoder.down_path.4.lin.weight", "encoder.down_path.5.lin.weight", "encoder.down_path.6.lin.weight", "decoder.up_path.0.lin.weight", "decoder.up_path.1.lin.weight", "decoder.up_path.2.lin.weight", "decoder.up_path.3.lin.weight", "decoder.up_path.4.lin.weight", "decoder.up_path.5.lin.weight", "decoder.up_path.6.lin.weight". 
	Unexpected key(s) in state_dict: "encoder.down_path.0.root", "encoder.down_path.1.root", "encoder.down_path.2.root", "encoder.down_path.3.root", "encoder.down_path.4.root", "encoder.down_path.5.root", "encoder.down_path.6.root", "decoder.up_path.0.root", "decoder.up_path.1.root", "decoder.up_path.2.root", "decoder.up_path.3.root", "decoder.up_path.4.root", "decoder.up_path.5.root", "decoder.up_path.6.root". 
```

When I print ```name, param in model.named_parameters():```, the two machines gave me different names. 
In the old version, it prints like this :
```encoder.down_path.0.weight torch.Size([125, 1, 8])
1000
encoder.down_path.0.root torch.Size([1, 8])
1008
encoder.down_path.0.bias torch.Size([8])
1016
...
```
In the new version, it prints like this :
```
encoder.down_path.0.weight torch.Size([125, 1, 8])
1000
encoder.down_path.0.bias torch.Size([8])
1008
encoder.down_path.0.lin.weight torch.Size([8, 1])
1016
```

How can I solve this issue, and why were their names changed?
Should I re-install the whole packages with specific versions installed on the old machine?

**Comment by zcaicaros** [comment]
I have had the same problem before. The way I solved it is to install the legacy version of[ PyG 1.7.2](https://github.com/pyg-team/pytorch_geometric/tree/1.7.2). Suppose you are using pytorch 1.7.0 + cuda 11.0, This is simple:
```
pip install --user --upgrade pip
pip install --user torch-scatter -f https://pytorch-geometric.com/whl/torch-1.7.0+cu110.html
pip install --user torch-sparse -f https://pytorch-geometric.com/whl/torch-1.7.0+cu110.html
# For additional but optional functionality, can be skipped:
pip install --user torch-cluster -f https://pytorch-geometric.com/whl/torch-1.7.0+cu110.html
# For additional but optional functionality, can be skipped:
pip install --user torch-spline-conv -f https://pytorch-geometric.com/whl/torch-1.7.0+cu110.html  
pip install --user torch-geometric==1.7.2
```

I guess in 2.0.2, they made some changes to the architectures (keys) of the network so when loading an old version of the model, the new interpreter cannot recognize the module's name.

**Reply by rusty1s** [reply]
Yes, one needs to sadly modify the names of the `state_dict`, see my answer below.

**Comment by rusty1s** [comment, marked as answer]
This is the result of our efforts to bring lazy initialization of model parameters to PyG. A simple fix is to replace all `root` occurrences with `lin.weight` and transpose the tensor, e.g.:
```python
state_dict["encoder.down_path.0.lin.weight"] = state_dict["encoder.down_path.0.root"].t()
```
